cdap-security/src/main/java/io/cdap/cdap/security/authorization/AuthorizerWrapper.java

```diff
@@ -0,0 +1,213 @@
+  public void destroy() {
+    try {
+      authorizer.destroy();
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  public Set<GrantedPermission> listGrants(Principal principal) throws AccessException {
+    try {
+      return authorizer.listPrivileges(principal).stream().map(
+        p -> new GrantedPermission(p.getAuthorizable(), p.getAction().getPermission())
+      ).collect(Collectors.toSet());
+    } catch (Exception e) {
+      throw AuthEnforceUtil.propagateAccessException(e);
+    }
+  }
+
+  @Override
+  public void grant(Authorizable authorizable, Principal principal, Set<? extends Permission> permissions)
+    throws AccessException {
+    try {
+      authorizer.grant(authorizable, principal, getActionSet(permissions));
+    } catch (Exception e) {
+      throw AuthEnforceUtil.propagateAccessException(e);
+    }
+  }
+
+  @Override
+  public void revoke(Authorizable authorizable, Principal principal, Set<? extends Permission> permissions)
+    throws AccessException {
+    try {
+      authorizer.revoke(authorizable, principal, getActionSet(permissions));
+    } catch (Exception e) {
+      throw AuthEnforceUtil.propagateAccessException(e);
+    }
+  }
+
+  @Override
+  public void revoke(Authorizable authorizable) throws AccessException {
+    try {
+      authorizer.revoke(authorizable);
+    } catch (Exception e) {
+      throw AuthEnforceUtil.propagateAccessException(e);
+    }
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Permission permission) throws AccessException {
+    try {
+      authorizer.enforce(entity, principal, getAction(permission));
+    } catch (Exception e) {
+      throw AuthEnforceUtil.propagateAccessException(e);
+    }
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Set<? extends Permission> permissions)
+    throws AccessException {
+    try {
+      authorizer.enforce(entity, principal, getActionSet(permissions));
+    } catch (Exception e) {
+      throw AuthEnforceUtil.propagateAccessException(e);
+    }
+  }
+
+  @Override
+  public void isVisible(EntityId entityId, Principal principal) throws AccessException {
+    try {
+      authorizer.isVisible(entityId, principal);
+    } catch (Exception e) {
+      throw AuthEnforceUtil.propagateAccessException(e);
+    }
+  }
+
+  @Override
+  public Set<? extends EntityId> isVisible(Set<? extends EntityId> entityIds, Principal principal)
+    throws AccessException {
+    try {
+      return authorizer.isVisible(entityIds, principal);
+    } catch (Exception e) {
+      throw AuthEnforceUtil.propagateAccessException(e);
+    }
+  }
+
+  private Set<Action> getActionSet(Set<? extends Permission> permissions) {
+    return permissions.stream().map(AuthorizerWrapper::getAction).collect(Collectors.toCollection(LinkedHashSet::new));
+  }
+
+  static Action getAction(Permission permission) {
+    if (StandardPermission.GET == permission) {
```

Comment:
As discussed offline, this architecture will retain the interface compatibility but will break backwards compatibility for the actual permissions checks; for example, a check which used to check ADMIN which now becomes StandardPermission.GET when converted back to an action will now become Action.READ which is incompatible with the original permissions model (even if it makes more semantic sense).

I believe the READ/WRITE actions only apply to the Dataset entities.

Reply:
Changed for the datasets